### PR TITLE
Refactor Models to Fully Utilize Allocators and Remove Bounds Dependency

### PR DESCRIPTION
**Pull request by yalsaffar**
#### PR Description:
This PR builds upon the changes introduced in #467 by fully removing the usage of lower and upper bounds within all models. 

## Key updates include:

- **Removal of Bounds and Dimension from Models:**  
  All `lb`, `ub`, and `dim` parameters have been fully removed from all models. The responsibility for computing inducing points is now entirely managed by allocators, ensuring a cleaner and more modular design.

- **Dependency Shift to Allocators:**  
  All model dependencies on dimensions (e.g., for computing the `covar_module`) are now managed through the `dim` property of the allocator object. This centralizes the handling of bounds and dimensions in the allocator system.

- **Updated Allocators Functionality:**  
  Allocators like `AutoAllocator` and `KMeansAllocator` now optionally take bounds. Bounds are essential in scenarios where no input data is provided, as they ensure fallback to the `DummyAllocator` (e.g., during initialization from configuration).

- **Tests:**  
  - All tests have been updated to make use of the allocator object as a requirement.  
  - Most tests pass bounds to the allocator, while others initialize bounds via configuration.  
  - Tests now fully align with the updated model-allocator interaction and function as expected.

Let me know if any changes are needed! :)

### CODE REVIEW

**Comment by yalsaffar** [comment]
There’s one failing test case in `test_strategy` that we’ve discussed previously. It’s related to the changes removing bounds from `MonotonicRejectionGP` and transitioning to the new allocator structure.

The issue seems to come from this line:

```python
fmean, _ = self.strat.model.predict(
    self.strat.eval_grid, probability_space=True
)
```

The result of `torch.max(fmean) - torch.min(fmean)` consistently falls below 0.3.  

---

#### Observations:
- **Old Implementation:**
  - Inducing points were initialized with Sobol, and the set allocator (also Sobol) was used.
  - The difference exceeded the threshold. ~0.32

- **New Implementation:**
  - Setting `SobolAllocator ` consistently results in a difference between 0.12 and 0.2.
  - Using `AutoAllocator` (initialized with `DummyAllocator`) gives a difference between 0.2 and 0.24.
  - Even replicating the old initialization (`SobolAllocator ` followed by `AutoAllocator`) produces similar results.



I’ve reviewed all the changes but can’t pinpoint the root cause of this behavior. 
Let me know if you have any suggestions or insights!

File: aepsych/models/monotonic_projection_gp.py [hunk 1]
```diff
@@ -173,7 +163,7 @@ def posterior(
             # using numpy because torch doesn't support vectorized linspace,
             # pytorch/issues/61292
             grid: Union[np.ndarray, torch.Tensor] = np.linspace(
-                self.lb[dim],
+                self.inducing_points.min(dim=0).values,
```
Comment (JasonKChow) [inline aepsych/models/monotonic_projection_gp.py:166]: This doesn't seem equivalent to before.

File: aepsych/strategy.py [hunk 2]
```diff
@@ -332,9 +328,41 @@ def get_max(
...
-        return self.model.get_max(
-            constraints, probability_space=probability_space, max_time=max_time
+        return self.get_max_(
+            self.model,
+            constraints,
+            probability_space=probability_space,
+            max_time=max_time,
+        )
+
+    def get_max_(
```
Comment (JasonKChow) [inline aepsych/strategy.py:338]: The actual functions should go outside of the strategy class, these should utility functions from the model subdirectory that is imported and called by the corresponding function in the strategy class.

File: tests/acquisition/test_mi.py [hunk 3]
```diff
@@ -91,8 +94,14 @@ def test_1d_single_probit(self):
         np.random.seed(seed)
         n_init = 15
         n_opt = 20
-        lb = -4.0
-        ub = 4.0
+        lb = torch.tensor([-4.0])
+        ub = torch.tensor([4.0])
+        inducing_size = 10
+        bounds = torch.stack([lb, ub])
+        inducing_points = select_inducing_points(
```
Comment (JasonKChow) [inline tests/acquisition/test_mi.py:101]: Should be auto.

File: tests/models/test_gp_classification.py [hunk 4]
```diff
@@ -160,13 +158,12 @@ def test_1d_classification_pytorchopt(self):
...
+        bounds = torch.stack([torch.tensor([-3.0]), torch.tensor([3.0])])
+
         model = GPClassificationModel(
-            torch.Tensor([-3]),
-            torch.Tensor([3]),
             inducing_size=10,
-            inducing_point_method=SobolAllocator(
-                bounds=torch.stack((torch.tensor([-3]), torch.tensor([3])))
-            ),
+            inducing_point_method=SobolAllocator(bounds=bounds),
```
Comment (JasonKChow) [inline tests/models/test_gp_classification.py:166]: Should be auto?

File: tests/models/test_gp_classification.py [hunk 5]
```diff
@@ -375,15 +371,11 @@ def test_1d_single_probit_new_interface(self):
...
                 ub=ub,
                 model=GPClassificationModel(
-                    lb=lb,
-                    ub=ub,
-                    inducing_size=10,
-                    inducing_point_method=SobolAllocator(
-                        bounds=torch.stack((torch.tensor([-4]), torch.tensor([4])))
-                    ),
+                    inducing_size=inducing_size,
+                    inducing_point_method=SobolAllocator(bounds=bounds),
```
Comment (JasonKChow) [inline tests/models/test_gp_classification.py:375]: Auto allocator.

File: tests/models/test_gp_classification.py [hunk 6]
```diff
@@ -428,15 +422,11 @@ def test_1d_single_probit_batched(self):
...
```
Comment (JasonKChow) [inline tests/models/test_gp_classification.py:426]: Auto allocators

File: tests/models/test_gp_classification.py [hunk 7]
```diff
@@ -480,15 +472,11 @@ def test_1d_single_probit(self):
...
```
Comment (JasonKChow) [inline tests/models/test_gp_classification.py:476]: Autoallocator

File: tests/models/test_gp_classification.py [hunk 8]
```diff
@@ -531,15 +521,11 @@ def test_1d_single_probit_pure_exploration(self):
...
```
Comment (JasonKChow) [inline tests/models/test_gp_classification.py:525]: Autoallocator.

File: tests/models/test_gp_classification.py [hunk 9]
```diff
@@ -587,17 +575,11 @@ def test_2d_single_probit_pure_exploration(self):
...
-                    lb=lb,
-                    ub=ub,
-                    inducing_size=10,
-                    inducing_point_method=SobolAllocator(
-                        bounds=torch.stack(
-                            (torch.tensor([-1, -1]), torch.tensor([1, 1]))
-                        )
-                    ),
+                    inducing_size=inducing_size,
+                    inducing_point_method=SobolAllocator(bounds=bounds),
```
Comment (JasonKChow) [inline tests/models/test_gp_classification.py:579]: Autoallocator.

File: tests/models/test_gp_classification.py [hunk 10]
```diff
@@ -648,15 +632,11 @@ def obj(x):
...
                 ub=ub,
                 model=GPClassificationModel(
-                    lb=lb,
-                    ub=ub,
-                    inducing_size=10,
-                    inducing_point_method=SobolAllocator(
-                        bounds=torch.stack((torch.tensor([-4]), torch.tensor([4])))
-                    ),
+                    inducing_size=inducing_size,
+                    inducing_point_method=SobolAllocator(bounds=bounds),
```
Comment (JasonKChow) [inline tests/models/test_gp_classification.py:636]: Autoallocator.

File: tests/models/test_gp_classification.py [hunk 11]
```diff
@@ -706,15 +688,11 @@ def obj(x):
...
```
Comment (JasonKChow) [inline tests/models/test_gp_classification.py:692]: Autoallocator.

File: tests/models/test_gp_classification.py [hunk 12]
```diff
@@ -770,16 +750,12 @@ def test_1d_single_lse(self):
...
```
Comment (JasonKChow) [inline tests/models/test_gp_classification.py:754]: Autoallocator.

File: tests/models/test_gp_classification.py [hunk 13]
```diff
@@ -872,15 +848,11 @@ def test_extra_ask_warns(self):
...
```
Comment (JasonKChow) [inline tests/models/test_gp_classification.py:852]: Autoallocator.

File: tests/models/test_gp_regression.py [hunk 14]
```diff
@@ -88,9 +90,9 @@ def test_extremum(self):
 
     def test_from_config(self):
         model = self.server.strat.model
-        npt.assert_allclose(model.transforms.untransform(model.lb), [-1.0])
-        npt.assert_allclose(model.transforms.untransform(model.ub), [3.0])
-        self.assertEqual(model.dim, 1)
+        # npt.assert_allclose(model.transforms.untransform(model.lb), [-1.0])
```
Comment (JasonKChow) [inline tests/models/test_gp_regression.py:93]: Reinstate these, check within the generator.

File: tests/models/test_monotonic_projection_gp.py [hunk 15]
```diff
@@ -54,6 +54,7 @@ def test_posterior(self):
         [MonotonicProjectionGP]
         monotonic_dims = [0, 1]
         inducing_size=10
+        dim = 3
```
Comment (JasonKChow) [inline tests/models/test_monotonic_projection_gp.py:57]: Why is this added?

File: tests/models/test_monotonic_projection_gp.py [hunk 16]
```diff
@@ -95,6 +96,7 @@ def test_posterior(self):
         monotonic_dims = [0]
         inducing_size=10
         min_f_val = 5.0
+        dim = 3
```
Comment (JasonKChow) [inline tests/models/test_monotonic_projection_gp.py:99]: See above.

File: tests/models/test_monotonic_projection_gp.py [hunk 17]
```diff
@@ -128,6 +130,7 @@ def test_from_config(self):
         monotonic_dims = [0]
         monotonic_grid_size = 10
         min_f_val = 0.1
+        dim = 2
```
Comment (JasonKChow) [inline tests/models/test_monotonic_projection_gp.py:133]: See above.

File: tests/models/test_semi_p.py [hunk 18]
```diff
@@ -26,51 +26,37 @@
...
     return HadamardSemiPModel(
-        lb=lb,
-        ub=ub,
         stim_dim=stim_dim,
         likelihood=BernoulliObjectiveLikelihood(objective=objective(floor=floor)),
         inducing_size=10,
-        inducing_point_method=AutoAllocator(
-            bounds=torch.stack((torch.tensor(lb), torch.tensor(ub)))
-        ),
+        inducing_point_method=inducing_point_method,
```
Comment (JasonKChow) [inline tests/models/test_semi_p.py:47]: This isn't equivalent, the AutoAllocator in inducing_point_method isn't initialized so this violates the model init spec.

File: tests/models/test_semi_p.py [hunk 19]
```diff
@@ -26,51 +26,37 @@
...
     return SemiParametricGPModel(
-        lb=lb,
-        ub=ub,
         stim_dim=stim_dim,
         likelihood=LinearBernoulliLikelihood(objective=objective(floor=floor)),
         inducing_size=10,
-        inducing_point_method=AutoAllocator(
-            bounds=torch.stack((torch.tensor(lb), torch.tensor(ub)))
-        ),
+        inducing_point_method=inducing_point_method,
```
Comment (JasonKChow) [inline tests/models/test_semi_p.py:59]: See above.

File: tests/server/test_server.py [hunk 20]
```diff
@@ -41,6 +41,7 @@
 [GPClassificationModel]
 inducing_size = 10
 mean_covar_factory = default_mean_covar_factory
+dim = 1
```
Comment (JasonKChow) [inline tests/server/test_server.py:44]: Why is this added?

File: tests/test_config.py [hunk 21]
```diff
@@ -124,19 +124,7 @@ def test_single_probit_config(self):
         self.assertTrue(strat.strat_list[0].outcome_types == ["binary"])
         self.assertTrue(strat.strat_list[1].min_asks == 20)
         self.assertTrue(torch.all(strat.strat_list[0].lb == strat.strat_list[1].lb))
-        self.assertTrue(
-            torch.all(
-                strat.transforms.untransform(strat.strat_list[1].model.lb)
-                == torch.Tensor([1, -1])
-            )
```
Comment (JasonKChow) [inline tests/test_config.py:131]: Check bounds, but check from the generator instead.

File: tests/test_config.py [hunk 22]
```diff
@@ -180,9 +169,7 @@ def test_single_probit_config_file(self):
         self.assertTrue(strat.strat_list[0].outcome_types == ["binary"])
         self.assertTrue(strat.strat_list[1].min_asks == 20)
         self.assertTrue(torch.all(strat.strat_list[0].lb == strat.strat_list[1].lb))
-        self.assertTrue(torch.all(strat.strat_list[1].model.lb == torch.Tensor([0, 0])))
```
Comment (JasonKChow) [inline tests/test_config.py:183]: Check bounds from generator instead.

File: tests/test_config.py [hunk 23]
```diff
@@ -216,9 +203,7 @@ def test_nonmonotonic_optimization_config_file(self):
...
```
Comment (JasonKChow) [inline tests/test_config.py:219]: Check bounds from generator instead.

File: tests/test_config.py [hunk 24]
```diff
@@ -519,9 +504,7 @@ def test_pairwise_probit_config(self):
...
```
Comment (JasonKChow) [inline tests/test_config.py:522]: Check bounds from generator instead.

File: tests/test_config.py [hunk 25]
```diff
@@ -557,9 +540,7 @@ def test_pairwise_probit_config_file(self):
...
```
Comment (JasonKChow) [inline tests/test_config.py:560]: Check bounds from generator instead.

File: tests/test_config.py [hunk 26]
```diff
@@ -598,9 +579,7 @@ def test_pairwise_al_config_file(self):
...
```
Comment (JasonKChow) [inline tests/test_config.py:601]: Check bounds from generator instead.

File: tests/test_config.py [hunk 27]
```diff
@@ -640,9 +619,7 @@ def test_pairwise_opt_config(self):
...
```
Comment (JasonKChow) [inline tests/test_config.py:643]: Check bounds from generator instead.

File: tests/test_config.py [hunk 28]
```diff
@@ -812,6 +789,9 @@ def test_stimuli_compatibility(self):
             [init_strat]
             generator = SobolGenerator
             model = PairwiseProbitModel
+
+            [PairwiseProbitModel]
+            dim = 2
```
Comment (JasonKChow) [inline tests/test_config.py:794]: Why is this added?

File: tests/test_config.py [hunk 29]
```diff
@@ -896,6 +876,10 @@ def test_outcome_compatibility(self):
             [init_strat]
             generator = SobolGenerator
             model = GPRegressionModel
+
+            [GPRegressionModel]
+            dim = 1
```
Comment (JasonKChow) [inline tests/test_config.py:881]: Why is dim added?

File: tests/test_config.py [hunk 30]
```diff
@@ -1004,6 +988,7 @@ def test_semip_config(self):
             inducing_size = 10
             inducing_point_method = SobolAllocator
             likelihood = BernoulliObjectiveLikelihood
+            dim = 2
```
Comment (JasonKChow) [inline tests/test_config.py:991]: Why dim?

File: tests/test_config.py [hunk 31]
```diff
@@ -1025,9 +1010,6 @@ def test_semip_config(self):
         model = opt_strat.model
 
         self.assertTrue(isinstance(model, HadamardSemiPModel))
-        self.assertTrue(torch.all(model.lb == torch.Tensor([0, 0])))
```
Comment (JasonKChow) [inline tests/test_config.py:1028]: Check bounds from generator.

File: tests/test_strategy.py [hunk 32]
```diff
@@ -112,9 +113,9 @@ def test_warmstart(self):
             self.strat.add_data(np.r_[1.0, 1.0], [1])
 
         self.assertEqual(
-            self.strat.model.fit.call_count, 4
+            self.strat.model.fit.call_count, 5
```
Comment (JasonKChow) [inline tests/test_strategy.py:116]: What has this changed?

File: tests/test_transforms.py [hunk 33]
```diff
@@ -92,17 +92,13 @@ def test_model_init_equivalent(self):
...
-            lb=torch.tensor([1, 1]),
-            ub=torch.tensor([100, 100]),
             inducing_point_method=AutoAllocator(
                 bounds=torch.stack((torch.tensor([1, 1]), torch.tensor([100, 100])))
             ),
             transforms=self.strat.strat_list[1].transforms,
         )
 
         self.assertTrue(type(config_model._base_obj) is type(obj_model._base_obj))
-        self.assertTrue(torch.equal(config_model.bounds, obj_model.bounds))
```
Comment (JasonKChow) [inline tests/test_transforms.py:104]: Check allocator bounds instead.

File: tests_gpu/generators/test_optimize_acqf_generator.py [hunk 34]
```diff
@@ -49,16 +49,20 @@
 class TestOptimizeAcqfGenerator(unittest.TestCase):
     @parameterized.expand(acqfs)
     def test_gpu_smoketest(self, acqf, acqf_kwargs):
-        lb = torch.tensor([0])
-        ub = torch.tensor([1])
+        self.device = torch.device("cuda")
+        lb = torch.tensor([0.0], device=self.device)
```
Comment (JasonKChow) [inline tests_gpu/generators/test_optimize_acqf_generator.py:53]: If this is failing it is actually a problem, you shouldn't need to manually change the device before init, the tensors should always follow the model's device.

File: tests_gpu/models/test_gp_regression.py [hunk 35]
```diff
@@ -57,6 +57,7 @@ def setUp(self):
             [GPRegressionModel]
             likelihood = GaussianLikelihood
             max_fit_time = 1
+            inducing_point_method = SobolAllocator
```
Comment (JasonKChow) [inline tests_gpu/models/test_gp_regression.py:60]: Why SobolAllocator?

**Review by JasonKChow — CHANGES_REQUESTED** [review]

File: tests/test_strategy.py (hunk 32 above)
Comment (yalsaffar) [reply tests/test_strategy.py:116]: Because now we are fitting the first time checking for `_is_first_fit` in Strategy.

File: aepsych/models/monotonic_projection_gp.py (hunk 1 above)
Comment (yalsaffar) [reply aepsych/models/monotonic_projection_gp.py:166]: Solved with using the lb and dim from inducing points method

File: aepsych/strategy.py (hunk 2 above)
Comment (yalsaffar) [reply aepsych/strategy.py:338]: I agree, these used to be methods inside `models/base.py` and I initially just moved them to `Strategy`, but I think there is no need to have them as we can directly call the functions in `models/utils.py`, which simplifies the whole structure. To do so, I had to re-implement `get_jnd() ` there.

File: tests_gpu/generators/test_optimize_acqf_generator.py (hunk 34 above)
Comment (yalsaffar) [reply tests_gpu/generators/test_optimize_acqf_generator.py:53]: That is true, but it does not apply to `OptimizeAcqfGenerator` as it requires bounds now, so I passed the device to it for now.

**Comment by JasonKChow** [comment]
Commandeered with #475
